### PR TITLE
Add support for fused Q Up-Proj GEMM/RoPE/Quant

### PR DESCRIPTION
**Pull request by chaseblock**
This commit adds support for fusing the q up proj gemm with the following rope and mxfp8 quantization operations. Relies on the TE functionality from [here](https://github.com/NVIDIA/TransformerEngine/pull/3303).

- [X] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Allows us to fuse the DeepseekV3 Q up proj gemm with the succeeding rope and mxfp8 quantization operations, improving training performance.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/transformer/multi_latent_attention.py [hunk 1]
```diff
@@ -851,17 +1074,61 @@ def qkv_up_proj_and_rope_apply(q_compressed, kv_compressed, k_pos_emb, rotary_po
...
                 q, _ = self.linear_q_up_proj(q_compressed)
+                import os as _os
+                if _os.environ.get("NVTE_FUSED_MLA_DEBUG", "0") == "1" and not getattr(self, '_unfused_dbg_done', False):
+                    self._unfused_dbg_done = True
+                    import torch.distributed as _dist
+                    if _dist.is_initialized() and _dist.get_rank() == 0:
+                        lin = self.linear_q_up_proj
+                        wq = lin._get_weight_quantizers()[0] if hasattr(lin, '_get_weight_quantizers') else None
+                        print(f"[UNFUSED_DBG] fp8={lin.fp8} weight_dtype={lin.weight.dtype} weight_type={type(lin.weight).__name__} weight_quantizer={type(wq).__name__ if wq else None}", flush=True)
+                # q: [num_tokens, n, q_head_dim]
```
Comment (sudhakarsingh27) [inline megatron/core/transformer/multi_latent_attention.py]: is this debug supposed to be in here?

Comment (chaseblock) [reply megatron/core/transformer/multi_latent_attention.py]: No it is not. Fixed.

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/transformer/multi_latent_attention.py [hunk 2]
```diff
@@ -52,36 +54,272 @@
...
     )
 except:
     fused_apply_mla_rope_for_kv = None
     fused_apply_mla_rope_for_q = None
+    rotary_bwd_q_kernel = None
 
 
 if HAVE_TE:
+    import transformer_engine_torch as tex
+    from transformer_engine.pytorch.attention import FusedMLAQUpProjRopeQuant
```
Comment (jiemingz) [inline megatron/core/transformer/multi_latent_attention.py]: should this be version checked? or else any old version of TE will do HAVE_TE=False

File: megatron/core/transformer/multi_latent_attention.py [hunk 3]
```diff
@@ -964,9 +1270,10 @@ def qkv_up_proj_and_rope_apply(q_compressed, kv_compressed, k_pos_emb, rotary_po
                     k_pos_emb = k_pos_emb.expand(-1, self.num_attention_heads_per_partition, -1)
                 key = torch.cat([k_no_pe, k_pos_emb], dim=-1)
 
-            query = query.contiguous()
-            key = key.contiguous()
-            value = value.contiguous()
+            if not use_fused_q_uproj:
```
Comment (jiemingz) [inline megatron/core/transformer/multi_latent_attention.py:1099]: why would the fused uproj need contiguous tensors here?

File: megatron/core/transformer/multi_latent_attention.py [hunk 4]
```diff
@@ -851,17 +1118,51 @@ def qkv_up_proj_and_rope_apply(q_compressed, kv_compressed, k_pos_emb, rotary_po
...
+                    b,
+                    self.tp_group,
+                    self.config.sequence_parallel,
+                )
+            elif self.config.q_lora_rank is not None:
                 # q_compressed: [num_tokens, q_lora_rank]
                 # q: [num_tokens, n * (qk_head_dim + qk_pos_emb_head_dim)]
                 q, _ = self.linear_q_up_proj(q_compressed)
+                # q: [num_tokens, n, q_head_dim]
+                q = q.view(*q.size()[:-1], self.num_attention_heads_per_partition, self.q_head_dim)
```
Comment (jiemingz) [inline megatron/core/transformer/multi_latent_attention.py]: nit: instead of duplicating this code can we just put them in the same else block

File: megatron/core/transformer/multi_latent_attention.py [hunk 5]
```diff
@@ -52,36 +54,272 @@
...
+        """Backward is unfused and matches the typical backward pass"""
+        x_saved, w_q, cos, sin = ctx.saved_tensors
+        nh, q_head_dim, qk_head_dim, qk_pos_emb_head_dim, s, b = ctx.dims
+        tokens = s * b
+        act_dtype = ctx.act_dtype
+
+        # --- RoPE backward (unchanged: bf16, same rotary_bwd_q_kernel as the unfused path) ---
+        dq3 = dq.reshape(tokens, nh, q_head_dim).contiguous()
+        grid = lambda META: (tokens, triton.cdiv(nh, META["BLOCK_H"]))
+        rotary_bwd_q_kernel[grid](
```
Comment (jiemingz) [inline megatron/core/transformer/multi_latent_attention.py]: are there plans to add a backward fusion? If not is there a way to avoid duplicating the backward pass that done here?

File: megatron/core/transformer/multi_latent_attention.py (hunk 3 above)
Comment (chaseblock) [reply megatron/core/transformer/multi_latent_attention.py:1099]: The fused path doesn't; the unfused does. We only added the guard here.

File: megatron/core/transformer/multi_latent_attention.py (hunk 2 above)
Comment (chaseblock) [reply megatron/core/transformer/multi_latent_attention.py]: This now handles import errors in the same way as the imports directly above it.

File: megatron/core/transformer/multi_latent_attention.py (hunk 4 above)
Comment (chaseblock) [reply megatron/core/transformer/multi_latent_attention.py]: Deduplicated.

File: megatron/core/transformer/multi_latent_attention.py (hunk 5 above)
Comment (chaseblock) [reply megatron/core/transformer/multi_latent_attention.py]: We haven't been considering a backwards fusion as far as I know, but I don't know that there's a better way to deduplicate this code at the moment.

I think this would require heavier changes to TE fully deduplicate this, and I'm not sure that this is worth it at this point.

File: megatron/core/transformer/multi_latent_attention.py (hunk 3 above)
Comment (jiemingz) [reply megatron/core/transformer/multi_latent_attention.py:1099]: sorry I meant *not need*, if the fused up_proj returns contiguous tensors, I think we can remove the if statement?

Comment (chaseblock) [reply megatron/core/transformer/multi_latent_attention.py:1099]: I think this .contiguous call would result in dequantizing the tensors.

In the unfused path, they aren't mxfp8 yet.

**Comment by sudhakarsingh27** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6213/commits/7c79c293d981872457b452ac4c042e24d0253c49

**Review by jiemingz — APPROVED** [review]

**Review by sudhakarsingh27 — APPROVED** [review]

**Comment by Phlip79** [comment]
I have updated tests to run on GB200 and not H100.
